### PR TITLE
Establish Default,Classic,Orange,Citrus Applab themes

### PR DESCRIPTION
**Pull request by cpirich**
* Mark has provided redlines for 13 Applab themes. These are the first 3 (Default, Orange, Citrus) plus the Classic theme that matches how Applab looked in the past. New screens will use 'Default'. Screens from existing projects will be set to 'Classic' when run in the `applabThemes` experiment for the first time.
* Added `fontFamily` and `fontSize` in themes for text area, text input, button, label, and dropdown
* Slightly modified the sizing logic when creating a button to more closely resemble the design comps (buttons created with themes that call for larger font sizes will be larger by default). Unlike the label sizing logic, which precisely fits to the text size, the button sizing logic is an estimate for when the initial button is dragged out (with the title `Button`) and it is always grid aligned

applab-four-themes

### CODE REVIEW

**Comment by cpirich** [comment]
@ryansloan FYI - I'll be on vacation next week, but you can see the results of the first themes here

File: apps/src/applab/constants.js [hunk 1]
```diff
@@ -18,9 +18,12 @@ export const CAPTURE_TICK_COUNT = 300;
 
 export const defaultFontSizeStyle = '14px';
 
-export const themeOptions = ['classic', 'dark'];
+export const DEFAULT_THEME_INDEX = 0;
+export const CLASSIC_THEME_INDEX = 1;
 
-export const themeDisplayNames = ['Classic', 'Dark'];
+export const themeOptions = ['default', 'classic', 'orange', 'citrus'];
```
Comment (jmkulwik) [inline apps/src/applab/constants.js:24]: Are we getting rid of dark theme?

**Review by jmkulwik — APPROVED** [review]

**Comment by joshlory** [comment]
Link to previous PR: https://github.com/code-dot-org/code-dot-org/pull/27644 (re-merged in https://github.com/code-dot-org/code-dot-org/pull/27672).

File: apps/src/applab/designElements/button.jsx [hunk 2]
```diff
@@ -208,41 +209,70 @@ export default {
...
       classic: color.white,
-      dark: color.black
+      orange: color.applab_orange_text_color,
+      citrus: color.applab_citrus_text_color
+    },
+    fontFamily: {
+      default: 'Arial Black',
+      classic: 'Arial',
+      orange: 'Verdana',
+      citrus: 'Georgia'
```
Comment (joshlory) [inline apps/src/applab/designElements/button.jsx:247]: Is there a way to track the theme styles in one place, instead of sprinkling the styles throughout the different components?

Related: intentional that orange uses 'Verdana' for buttons, but 'Arial' for labels?

File: shared/css/color.scss [hunk 3]
```diff
@@ -118,5 +118,47 @@ $droplet_purple: #bb77c7;
 $droplet_green: #68d995;
 $droplet_white: $white;
 
+// Applab theme values.
```
Comment (joshlory) [inline shared/css/color.scss:121]: I'm not sure it's super useful for these to be in the global color.scss unless you think they'll get used outside of App Lab.

File: apps/src/applab/gridUtils.js [hunk 4]
```diff
@@ -101,4 +101,14 @@ export function snapToGridSize(coordinate) {
...
+/**
+ * Given a dimension (width or height), returns a dimension of at least
+ * that size that aligns to the given grid size.
+ * @param {number} dimension
+ * @returns {number}
+ */
+export function growToGridSize(dimension) {
+  return Math.ceil(dimension / GRID_SIZE) * GRID_SIZE;
+}
+
```
Comment (joshlory) [inline apps/src/applab/gridUtils.js:113]: I'd love to see this functionality change pulled out into a separate PR!

File: apps/src/applab/commands.js [hunk 5]
```diff
@@ -1008,12 +1010,12 @@ applabCommands.dropdown = function(opts) {
...
   newSelect.id = opts.elementId;
   newSelect.style.position = 'relative';
-  newSelect.style.fontSize = defaultFontSizeStyle;
-  newSelect.style.fontFamily = fontFamilyStyles[0];
   if (experiments.isEnabled('applabThemes')) {
     newSelect.style.borderStyle = 'solid';
     elementLibrary.applyCurrentTheme(newSelect, Applab.activeScreen());
   } else {
+    newSelect.style.fontSize = defaultFontSizeStyle;
+    newSelect.style.fontFamily = fontFamilyStyles[0];
```
Comment (joshlory) [inline apps/src/applab/commands.js:1018]: Echoing @islemaster's feedback on the original PR, can you unify this logic for the experiment and non-experiment case? https://github.com/code-dot-org/code-dot-org/pull/27644#discussion_r269662193

Unless you _really_ don't think we'll go this direction for shipping the code.

**Comment by islemaster** [comment]
I'm a little confused by "default" as the name of a theme.  It seems like all of our themes should have descriptive names, and while one of them might be the default for new projects at some point, that doesn't necessarily have to be its name.

Also, what's the rollout plan for this? Do all existing projects get "classic" theme and new projects get "default" theme?

File: apps/src/applab/designElements/button.jsx (hunk 2 above)
Comment (islemaster) [reply apps/src/applab/designElements/button.jsx:247]: Agreed, I'm a little worried about this logic being spread out across all of the App Lab design elements.  It seems like it increases the cost of adding a new design element or adding a new theme.  It's possible that having these four themes we are just starting to see the effect of that.  Can we have a more detailed discussion about what a different approach would look like?

File: apps/src/applab/designElements/button.jsx [hunk 6]
```diff
@@ -208,41 +209,70 @@ export default {
...
-    element.style.fontSize = defaultFontSizeStyle;
     if (experiments.isEnabled('applabThemes')) {
       element.style.borderStyle = 'solid';
+      // Roughly scale default size based on the current theme's font size:
+      const currentTheme = elementLibrary.getCurrentTheme(
+        designMode.activeScreen()
+      );
+      const fontSize = this.themeValues.fontSize[currentTheme];
+      element.style.height = `${growToGridSize(fontSize * 2)}px`;
+      element.style.width = `${10 + growToGridSize(fontSize * 5)}px`;
```
Comment (islemaster) [inline apps/src/applab/designElements/button.jsx:269]: Just making sure - this only happens when we create the element, not when changing themes, correct?  Does manually setting these styles mess with resize-as-you-type behavior for elements with default sizes at all?

File: apps/src/applab/commands.js [hunk 7]
```diff
@@ -199,12 +199,12 @@ applabCommands.button = function(opts) {
   var textNode = document.createTextNode(opts.text);
   newButton.id = opts.elementId;
   newButton.style.position = 'relative';
-  newButton.style.fontSize = defaultFontSizeStyle;
-  newButton.style.fontFamily = fontFamilyStyles[0];
```
Comment (islemaster) [inline apps/src/applab/commands.js:203]: Hm - was going to ask why we don't apply the special theme-specific sizing logic here, but might see it already.  It looks like these two button-creation paths (via design mode, via code) have never quite been unified and it might break existing apps to unify them now.  Default button size is already different for the two paths.

File: apps/src/applab/constants.js (hunk 1 above)
Comment (cpirich) [reply apps/src/applab/constants.js:24]: yes, dark theme was temporary from the @poorvasingal spec (no design input)

File: apps/src/applab/designElements/button.jsx (hunk 2 above)
Comment (cpirich) [reply apps/src/applab/designElements/button.jsx:247]: It is not hard to imagine a consolidation of theme values in a different file. It may be easier for maintenance in the long term. However, there is also a fair amount of code for each component that is interrelated to these themes (when to backfill certain properties with "classic" during deserialize, which properties should be set on creation, etc.)

It may be easier to consolidate after we remove the experiment flag and the older code that runs when the experiment is not enabled

File: shared/css/color.scss (hunk 3 above)
Comment (cpirich) [reply shared/css/color.scss:121]: Some of the default applab values were there, so I tried to keep them in the same place. I agree that we should look at moving these out.

File: apps/src/applab/commands.js (hunk 5 above)
Comment (cpirich) [reply apps/src/applab/commands.js:1018]: The goal is to flip on the experiment and delete the old code within the next 1-2 weeks. I'd rather preserve the existing code paths as much as possible for the non-experiment case until I delete it at the end of the month.

**Comment by cpirich** [comment]
> I'm a little confused by "default" as the name of a theme. It seems like all of our themes should have descriptive names, and while one of them might be the default for new projects at some point, that doesn't necessarily have to be its name.
> 
> Also, what's the rollout plan for this? Do all existing projects get "classic" theme and new projects get "default" theme?

@ryansloan should probably handle this response, but you are correct (as mentioned in the PR description), existing screens (created before applab themes) become "Classic". New screens are "Default".

File: apps/src/applab/designElements/button.jsx (hunk 6 above)
Comment (cpirich) [reply apps/src/applab/designElements/button.jsx:269]: This is only for the button. There is no resize-as-you-type behavior on that element. Previously, all buttons created in design mode were hardcoded as 80x30 pixels. That will still be the case if the screen is in the "Classic" theme. However, they may be larger if created when the screen is in one of the other new themes.

File: apps/src/applab/commands.js (hunk 7 above)
Comment (cpirich) [reply apps/src/applab/commands.js:203]: Correct, and I didn't want to change that. The 80x30 default size only applied to design mode created buttons.

File: apps/src/applab/designElements/button.jsx (hunk 2 above)
Comment (joshlory) [reply apps/src/applab/designElements/button.jsx:247]: @cpirich

> Related: intentional that orange uses 'Verdana' for buttons, but 'Arial' for labels?
